### ISSUE TITLE
babel-traverse and react-docgen

### ISSUE THREAD

**Issue by chrisui** [reactjs/react-docgen#68]
Hey! I'm doing some work with webpack, babel and react-docgen and I've hit a particular problem where I'm not really sure what the best solution is so looking for some advice.

At a glance I have a webpack loader which consumes a "component manifest" at build time and with that can resolve the appropriate component (following where the component was referenced from through imports) and generate documentation (conveniently just using docgen) for use later.

I have all this logic, to track dependencies and resolve components, which is written using babel-traverse on a tree generated by babylon and I was hoping I would simply be able to re-use this to pass into docgen.

However, it seems there are some subtle differences with the outputted `NodePath`s which mean docgen cannot extract information correctly (More accurately it errors due to bad references all over the shop). 

One example is when a `VariableDeclaration` is resolved the `node.declarations` is (by docgen with recast) expected to be an object with a `value` property however when resolved using babel-traverse it is actually, more simply, an array (interesting thing here was on ASTExplorer the outputted structure was the same - which makes me think I'm probably missing something trivial).

Is there something I'm missing with the paths provided by babel-traverse that would make them work with recast or are they fundamentally different api's meaning they would need some sort of conversion step?

Also, out of interest, wondering how the decision is made to use something like recast over babel-traverse (considering babylon is already used as the parser)? Guess this may be a timing thing when work was actually started on this! Also, is it maybe the plan to move to babel (making assumptions since it's FB managed now)?

In summary I'm looking to re-use a bunch of logic I've already written as babel based traversals and am desperately hoping I can take advantage of docgen in the least ugliest manner possible. :)

**Comment by fkling** [comment]
> interesting thing here was on ASTExplorer the outputted structure was the same - which makes me think I'm probably missing something trivial

The ASTs are probably the same, but the NodePath's may not?

> Is there something I'm missing with the paths provided by babel-traverse that would make them work with recast or are they fundamentally different api's meaning they would need some sort of conversion step?

I don't know that much about the NodePath implement babel-traverse. But from writing a custom Babel plugin I can say that they appear to be very different than recast's NodePaths.

> Also, out of interest, wondering how the decision is made to use something like recast over babel-traverse (considering babylon is already used as the parser)? Guess this may be a timing thing when work was actually started on this!

Pretty much, yes. I actually wrote jscodeshift two years ago. Babel didn't exist back than.

> Also, is it maybe the plan to move to babel (making assumptions since it's FB managed now)?

I don't have any plans in this direction. recast works well for us as it is and I believe that a conversion is worth the effort (not enough gain).

>  I'm looking to re-use a bunch of logic I've already written as babel based traversals and am desperately hoping I can take advantage of docgen in the least ugliest manner possible.

docgen uses babylon v5. If you are using v6, then that could explain some of the differences. But aside that, there shouldn't be an issue using a babylon AST with docgen/recast. I'm happy to provide an API that accepts an AST (instead of code).

Maybe if you provide concrete examples which cause issues, we can solve them?

**Comment by fkling** [comment]
I'm closing this since I guess there doesn't seem to be anything actionable here.
